rust/lance-index/src/vector/pq.rs

```diff
@@ -240,16 +241,63 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
                 distance_table.extend(distances);
             });
 
-        Ok(Arc::new(Float32Array::from_iter_values(
-            code.values().chunks_exact(self.num_sub_vectors).map(|c| {
-                c.iter()
-                    .enumerate()
-                    .map(|(sub_vec_idx, centroid)| {
-                        distance_table[sub_vec_idx * 256 + *centroid as usize]
-                    })
-                    .sum()
-            }),
-        )))
+        #[cfg(target_feature = "avx512f")]
+        {
+            if self.num_sub_vectors % 16 == 0 {
```

Comment:
can we do the first parts with avx512 and the smaller part with avx2 or sse?

Or does that not help with perf

Reply:
From tracing, this piece is major bottleneck as it is linear to # of rows in this partitions.